### ISSUE TITLE
Investigate to leverage accessibility API in Electron 

### ISSUE THREAD

**Issue by bpasero** [Microsoft/vscode#14592]
Since we are on Electron 1.3.x there is more API support to find out if a screen reader is attached or not. The related API is:
- `accessibility-support-changed` [event](https://github.com/electron/electron/blob/master/docs/api/app.md#event-accessibility-support-changed-macos-windows)
- a new [method](https://github.com/electron/electron/blob/master/docs/api/app.md#appisaccessibilitysupportenabled-macos-windows) `app.isAccessibilitySupportEnabled()`

I verified that these work on Windows and Mac. On Mac the event works when changing the native screen reader (Cmd+F5). On Windows I tried with NVDA and the only thing that was not working is that the event would fire when turning OFF NVDA, but turning ON was ok.

@alexandrudima I think you expressed some interest in this, so I am moving this to you first to see if we want to use it. I can then look into how to surface this API as workbench API for you.

**Comment by alexandrudima** [comment]
I would like to use it, but I remember reading in an Electron issue that the event is not always fired (i.e. when Chromium decides to go in accessibility mode vs. when Electron decides to put Chromium in accessibility mode)

**Comment by alexandrudima** [comment]
Also, for me, the fix for entering accessibility mode too eagerly does not work.

**Comment by bpasero** [comment]
I did some simple testing with VoiceOver and NVDA and it seemed to do the trick. Of course I am not sure about the corner cases. Maybe we could do a combination of both: detect a11y mode and provide a setting to force enable if the detection does not work properly for some reason.

Can you clarify what you mean by https://github.com/Microsoft/vscode/issues/14592#issuecomment-256641685 ?

**Comment by alexandrudima** [comment]
I mean I need to patch the Code - Insiders.exe every day because it enters accessibility mode.

**Comment by bpasero** [comment]
@alexandrudima maybe we are not talking about the same thing here. I was thinking that the editor could maybe turn into a less expensive mode when accessibility is detected to be OFF and turn into a more accessible mode once we detect a screen reader is enabled. Somehow I remember that you need to do quite some tricks to get something for the screen reader from the editor and maybe those tricks are not needed when a screen reader is disabled.

Now what you are talking about is the fact that on your machine (not my Thinkpad and also not Dirks Surface Book), Electron enables accessibility mode even though you are not running a screen reader. I can only suspect that you must be running some other software or tweaked some other setting in Windows that makes it behave different from Dirks machine.

Can you add this code to our main.ts to see what the API returns for you?

* `app.isAccessibilitySupportEnabled()` to find out the state right on startup
* app.on('accessibility-support-changed', function()...) to console.log when the a11y mode changes?

It would be interesting to run this with and without NVDA running.

**Comment by alexandrudima** [comment]
@bpasero This is what I'm talking about:

`app.on('accessibility-support-changed')` is not fired when Chromium brings itself (yes, it does that sometimes by itself!) into accessibility mode according to the comment -> https://github.com/electron/electron/pull/7611#issuecomment-253944947

i.e. the event is not fired when `BrowserAccessibilityState::IsAccessibleBrowser()` changes its value, it only fires when Electron decides to bring Chromium into accessibility mode.

We could workaround this by installing an interval and asking every x seconds if it entered accessibility mode.

Yes, when I finish reading my notifications, I will check out electron 1.3.8 and see what goes on on my machine.

**Comment by bpasero** [comment]
Ok thanks for the infos.

**Comment by alexandrudima** [comment]
I've got this almost working, I just need to add an emergency option to turn it on/off if Electron's detection mechanism fails and they don't fire the event.

e.g. Opening the Accessibility Options menu entry and enabling from there does not fire the event.

**Comment by alexandrudima** [comment]
In the Stable version, we don't ship with the Accessibility Options menu entry, so IMHO there is no way to enable Chromium accessibility such that Electron missed the event. Let's test this and see if there are any series of steps where Electron does not fire the event and add an option only if necessary.
